File: scripts/compare_checksums.py
#!/usr/bin/env python3

import os
import sys
import hashlib
import argparse
import json
from collections import defaultdict
from pathlib import Path
from typing import Dict, List, Set, Tuple

def calculate_file_checksum(file_path: str) -> str:
    """Calculate SHA-256 checksum of a file."""
    sha256_hash = hashlib.sha256()
    with open(file_path, "rb") as f:
        # Read and update hash in chunks of 4K
        for byte_block in iter(lambda: f.read(4096), b""):
            sha256_hash.update(byte_block)
    return sha256_hash.hexdigest()

def get_all_files(directory: str) -> List[str]:
    """Get all files in a directory recursively."""
    all_files = []
    for root, _, files in os.walk(directory):
        for file in files:
            all_files.append(os.path.join(root, file))
    return all_files

def get_relative_path(file_path: str, base_dir: str) -> str:
    """Get the path of a file relative to a base directory."""
    return os.path.relpath(file_path, base_dir)

def find_builds(dist_record_dir: str, version: str) -> List[str]:
    """Find all builds for a specific version."""
    version_dir = os.path.join(dist_record_dir, version)
    if not os.path.exists(version_dir):
        print(f"No builds found for version {version}")
        return []
    
    builds = []
    for item in os.listdir(version_dir):
        if item.startswith("build-"):
            builds.append(os.path.join(version_dir, item))
    
    return sorted(builds)

def compare_checksums(builds: List[str]) -> Tuple[Dict, Dict]:
    """Compare checksums across builds."""
    # Dictionary to store checksums for each file in each build
    # {relative_file_path: {build_number: checksum}}
    checksums = defaultdict(dict)
    
    # Dictionary to store file sizes for each file in each build
    # {relative_file_path: {build_number: size}}
    file_sizes = defaultdict(dict)
    
    # Process each build
    for build_path in builds:
        build_number = os.path.basename(build_path).replace("build-", "")
        
        # Process dist-firefox directory
        dist_firefox_dir = os.path.join(build_path, "dist-firefox")
        if os.path.exists(dist_firefox_dir):
            for file_path in get_all_files(dist_firefox_dir):
                rel_path = get_relative_path(file_path, build_path)
                checksums[rel_path][build_number] = calculate_file_checksum(file_path)
                file_sizes[rel_path][build_number] = os.path.getsize(file_path)
                
        # Process dist-chrome directory
        dist_chrome_dir = os.path.join(build_path, "dist-chrome")
        if os.path.exists(dist_chrome_dir):
            for file_path in get_all_files(dist_chrome_dir):
                rel_path = get_relative_path(file_path, build_path)
                checksums[rel_path][build_number] = calculate_file_checksum(file_path)
                file_sizes[rel_path][build_number] = os.path.getsize(file_path)
        
        # Process dist-zip directory
        dist_zip_dir = os.path.join(build_path, "dist-zip")
        if os.path.exists(dist_zip_dir):
            for file_path in get_all_files(dist_zip_dir):
                rel_path = get_relative_path(file_path, build_path)
                checksums[rel_path][build_number] = calculate_file_checksum(file_path)
                file_sizes[rel_path][build_number] = os.path.getsize(file_path)
    
    return checksums, file_sizes

def find_inconsistencies(checksums: Dict, file_sizes: Dict) -> Tuple[Dict, Dict, Dict]:
    """Find files with inconsistent checksums across builds."""
    inconsistent_files = {}
    missing_files = {}
    size_differences = {}
    
    # Precompute all build numbers
    all_build_numbers = set(build_number for checksums_dict in checksums.values() for build_number in checksums_dict.keys())
    
    for file_path, build_checksums in checksums.items():
        # Check if file exists in all builds
        build_numbers = list(build_checksums.keys())
        
        if len(build_numbers) < len(all_build_numbers):
            missing_in_builds = all_build_numbers - set(build_numbers)
            missing_files[file_path] = sorted(missing_in_builds)
        
        # Check if checksums are consistent across builds
        unique_checksums = set(build_checksums.values())
        if len(unique_checksums) > 1:
            inconsistent_files[file_path] = build_checksums
            
            # Check if size is different
            file_build_sizes = file_sizes[file_path]
            unique_sizes = set(file_build_sizes.values())
            if len(unique_sizes) > 1:
                size_differences[file_path] = file_build_sizes
    
    return inconsistent_files, missing_files, size_differences

def generate_report(
    inconsistent_files: Dict, 
    missing_files: Dict, 
    size_differences: Dict, 
    builds: List[str],
    output_file: str = None
) -> None:
    """Generate a report of inconsistencies."""
    build_numbers = [os.path.basename(build).replace("build-", "") for build in builds]
    
    report = {
        "builds_analyzed": build_numbers,
        "total_builds": len(build_numbers),
        "inconsistent_files": {
            "count": len(inconsistent_files),
            "files": {file_path: checksums for file_path, checksums in inconsistent_files.items()}
        },
        "missing_files": {
            "count": len(missing_files),
            "files": {file_path: missing_builds for file_path, missing_builds in missing_files.items()}
        },
        "size_differences": {
            "count": len(size_differences),
            "files": {file_path: sizes for file_path, sizes in size_differences.items()}
        }
    }
    
    # Print summary to console
    print(f"\nChecksum Comparison Report")
    print(f"========================")
    print(f"Builds analyzed: {', '.join(build_numbers)}")
    print(f"Total builds: {len(build_numbers)}")
    print(f"Files with inconsistent checksums: {len(inconsistent_files)}")
    print(f"Files missing in some builds: {len(missing_files)}")
    print(f"Files with size differences: {len(size_differences)}")
    
    # Print details of inconsistent files
    if inconsistent_files:
        print("\nInconsistent Files:")
        for file_path, checksums in inconsistent_files.items():
            print(f"\n  {file_path}:")
            for build, checksum in checksums.items():
                print(f"    Build {build}: {checksum}")
    
    # Write report to file if specified
    if output_file:
        with open(output_file, 'w') as f:
            json.dump(report, f, indent=2)
        print(f"\nDetailed report saved to {output_file}")

def main():
    parser = argparse.ArgumentParser(description="Compare checksums across builds of the same version")
    parser.add_argument("version", help="Version to compare builds for")
    parser.add_argument("--dist-record", default="../dist-record", help="Path to dist-record directory")
    parser.add_argument("--output", help="Output file for detailed report (JSON format)")
    
    args = parser.parse_args()
    
    # Resolve paths
    script_dir = os.path.dirname(os.path.abspath(__file__))
    dist_record_dir = os.path.abspath(os.path.join(script_dir, args.dist_record))
    
    # Find builds for the specified version
    builds = find_builds(dist_record_dir, args.version)
    if not builds:
        sys.exit(1)
    
    print(f"Found {len(builds)} builds for version {args.version}")
    for build in builds:
        print(f"  - {os.path.basename(build)}")
    
    # Compare checksums
    print("\nCalculating checksums...")
    checksums, file_sizes = compare_checksums(builds)
    
    # Find inconsistencies
    print("Finding inconsistencies...")
    inconsistent_files, missing_files, size_differences = find_inconsistencies(checksums, file_sizes)
    
    # Generate report
    output_file = args.output
    if output_file and not os.path.isabs(output_file):
        output_file = os.path.abspath(os.path.join(os.getcwd(), output_file))
    
    generate_report(inconsistent_files, missing_files, size_differences, builds, output_file)

if __name__ == "__main__":
    main()

File: scripts/compare-checksums.js
/**
 * @file compare-checksums.js
 * Compare checksums across builds of the same version to verify build reproducibility.
 * Usage: bun run compare-checksums [version] [--output report.json]
 */
import fs from "fs";
import path from "path";
import crypto from "crypto";
import { getCurrentVersion, ROOT_DIR } from "./lib/utils.js";

/**
 * Calculate SHA-256 checksum of a file.
 * @param {string} filePath - Path to the file.
 * @returns {string} Hex-encoded SHA-256 hash.
 */
function calculateFileChecksum(filePath) {
  const hash = crypto.createHash("sha256");
  const fileBuffer = fs.readFileSync(filePath);
  hash.update(fileBuffer);
  return hash.digest("hex");
}

/**
 * Get all files in a directory recursively.
 * @param {string} directory - Directory to scan.
 * @returns {string[]} Array of file paths.
 */
function getAllFiles(directory) {
  const files = [];

  function walkDir(dir) {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walkDir(fullPath);
      } else {
        files.push(fullPath);
      }
    }
  }

  walkDir(directory);
  return files;
}

/**
 * Find all builds for a specific version.
 * @param {string} distRecordDir - Path to dist-record directory.
 * @param {string} version - Version to find builds for.
 * @returns {string[]} Array of build directory paths.
 */
function findBuilds(distRecordDir, version) {
  const versionDir = path.join(distRecordDir, version);

  if (!fs.existsSync(versionDir)) {
    console.log(`No builds found for version ${version}`);
    return [];
  }

  const builds = fs
    .readdirSync(versionDir, { withFileTypes: true })
    .filter((entry) => entry.isDirectory() && entry.name.startsWith("build-"))
    .map((entry) => path.join(versionDir, entry.name))
    .sort();

  return builds;
}

/**
 * Compare checksums across builds.
 * @param {string[]} builds - Array of build directory paths.
 * @returns {{checksums: Object, fileSizes: Object}} Checksums and file sizes by file and build.
 */
function compareChecksums(builds) {
  // {relative_file_path: {build_number: checksum}}
  const checksums = {};
  // {relative_file_path: {build_number: size}}
  const fileSizes = {};

  for (const buildPath of builds) {
    const buildNumber = path.basename(buildPath).replace("build-", "");

    // Process browser-specific directories
    for (const dirname of ["dist-firefox", "dist-chrome"]) {
      const distDir = path.join(buildPath, dirname);
      if (fs.existsSync(distDir)) {
        for (const filePath of getAllFiles(distDir)) {
          const relPath = path.relative(buildPath, filePath);

          if (!checksums[relPath]) {
            checksums[relPath] = {};
            fileSizes[relPath] = {};
          }

          checksums[relPath][buildNumber] = calculateFileChecksum(filePath);
          fileSizes[relPath][buildNumber] = fs.statSync(filePath).size;
        }
      }
    }

    // Process dist-zip directory
    const distZipDir = path.join(buildPath, "dist-zip");
    if (fs.existsSync(distZipDir)) {
      for (const filePath of getAllFiles(distZipDir)) {
        const relPath = path.relative(buildPath, filePath);

        if (!checksums[relPath]) {
          checksums[relPath] = {};
          fileSizes[relPath] = {};
        }

        checksums[relPath][buildNumber] = calculateFileChecksum(filePath);
        fileSizes[relPath][buildNumber] = fs.statSync(filePath).size;
      }
    }
  }

  return { checksums, fileSizes };
}

/**
 * Find files with inconsistent checksums across builds.
 * @param {Object} checksums - Checksums by file and build.
 * @param {Object} fileSizes - File sizes by file and build.
 * @returns {{inconsistentFiles: Object, missingFiles: Object, sizeDifferences: Object}}
 */
function findInconsistencies(checksums, fileSizes) {
  const inconsistentFiles = {};
  const missingFiles = {};
  const sizeDifferences = {};

  // Get all build numbers
  const allBuildNumbers = new Set();
  for (const buildChecksums of Object.values(checksums)) {
    for (const buildNumber of Object.keys(buildChecksums)) {
      allBuildNumbers.add(buildNumber);
    }
  }

  for (const [filePath, buildChecksums] of Object.entries(checksums)) {
    const buildNumbers = Object.keys(buildChecksums);

    // Check if file exists in all builds
    if (buildNumbers.length < allBuildNumbers.size) {
      const missingInBuilds = [...allBuildNumbers].filter((b) => !buildNumbers.includes(b));
      missingFiles[filePath] = missingInBuilds.sort();
    }

    // Check if checksums are consistent across builds
    const uniqueChecksums = new Set(Object.values(buildChecksums));
    if (uniqueChecksums.size > 1) {
      inconsistentFiles[filePath] = buildChecksums;

      // Check if size is different
      const fileBuildSizes = fileSizes[filePath];
      const uniqueSizes = new Set(Object.values(fileBuildSizes));
      if (uniqueSizes.size > 1) {
        sizeDifferences[filePath] = fileBuildSizes;
      }
    }
  }

  return { inconsistentFiles, missingFiles, sizeDifferences };
}

/**
 * Generate and print a report of inconsistencies.
 * @param {Object} inconsistentFiles - Files with inconsistent checksums.
 * @param {Object} missingFiles - Files missing in some builds.
 * @param {Object} sizeDifferences - Files with size differences.
 * @param {string[]} builds - Build directory paths.
 * @param {string|null} outputFile - Optional output file for JSON report.
 */
function generateReport(inconsistentFiles, missingFiles, sizeDifferences, builds, outputFile) {
  const buildNumbers = builds.map((build) => path.basename(build).replace("build-", ""));

  const report = {
    builds_analyzed: buildNumbers,
    total_builds: buildNumbers.length,
    inconsistent_files: {
      count: Object.keys(inconsistentFiles).length,
      files: inconsistentFiles,
    },
    missing_files: {
      count: Object.keys(missingFiles).length,
      files: missingFiles,
    },
    size_differences: {
      count: Object.keys(sizeDifferences).length,
      files: sizeDifferences,
    },
  };

  // Print summary to console
  console.log("\nChecksum Comparison Report");
  console.log("========================");
  console.log(`Builds analyzed: ${buildNumbers.join(", ")}`);
  console.log(`Total builds: ${buildNumbers.length}`);
  console.log(`Files with inconsistent checksums: ${Object.keys(inconsistentFiles).length}`);
  console.log(`Files missing in some builds: ${Object.keys(missingFiles).length}`);
  console.log(`Files with size differences: ${Object.keys(sizeDifferences).length}`);

  // Print details of inconsistent files
  if (Object.keys(inconsistentFiles).length > 0) {
    console.log("\nInconsistent Files:");
    for (const [filePath, checksums] of Object.entries(inconsistentFiles)) {
      console.log(`\n  ${filePath}:`);
      for (const [build, checksum] of Object.entries(checksums)) {
        console.log(`    Build ${build}: ${checksum}`);
      }
    }
  }

  // Write report to file if specified
  if (outputFile) {
    const outputPath = path.isAbsolute(outputFile) ? outputFile : path.join(process.cwd(), outputFile);
    fs.writeFileSync(outputPath, JSON.stringify(report, null, 2));
    console.log(`\nDetailed report saved to ${outputPath}`);
  }

  // Return exit code based on inconsistencies
  return Object.keys(inconsistentFiles).length === 0 ? 0 : 1;
}

/**
 * Parse command line arguments.
 * @returns {{version: string, output: string|null}}
 */
function parseArgs() {
  const args = process.argv.slice(2);
  let version = null;
  let output = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--output" && args[i + 1]) {
      output = args[i + 1];
      i++;
    } else if (!args[i].startsWith("-")) {
      version = args[i];
    }
  }

  // If no version provided, use current version from package.json
  if (!version) {
    version = getCurrentVersion();
    console.log(`No version specified. Using current version from package.json: ${version}`);
  }

  return { version, output };
}

/**
 * Main function.
 */
function main() {
  const { version, output } = parseArgs();
  const distRecordDir = path.join(ROOT_DIR, "dist-record");

  // Find builds for the specified version
  const builds = findBuilds(distRecordDir, version);
  if (builds.length === 0) {
    process.exit(1);
  }

  console.log(`Found ${builds.length} builds for version ${version}`);
  for (const build of builds) {
    console.log(`  - ${path.basename(build)}`);
  }

  if (builds.length < 2) {
    console.log("\n⚠ Warning: Need at least 2 builds to compare. Run 'bun build:all --record' again.");
    process.exit(0);
  }

  // Compare checksums
  console.log("\nCalculating checksums...");
  const { checksums, fileSizes } = compareChecksums(builds);

  // Find inconsistencies
  console.log("Finding inconsistencies...");
  const { inconsistentFiles, missingFiles, sizeDifferences } = findInconsistencies(checksums, fileSizes);

  // Generate report
  const exitCode = generateReport(inconsistentFiles, missingFiles, sizeDifferences, builds, output);

  process.exit(exitCode);
}

main();
